source/client-side-encryption/client-side-encryption.rst

```diff
@@ -770,73 +942,299 @@ behind-the-scenes operations required for encryption or decryption.
+
+.. note::
+
+   The driver MUST NOT manipulate or do any validation on the csfle path options
+   provided in `cse.extraoptions`. They should be passed through to
+   libmongocrypt unchanged.
+
+
+Path Resolution Behavior
+------------------------
+
+.. |---| unicode:: U+2014
+
+Drivers should include and note the following information regarding the behavior
+of `csfle` path options in `cse.extraOptions`:
+
+- If used, the `override path`_ must be given as a path to the `csfle` dynamic
+  library file *itself*, and not simply the directory that contains it.
+
+- If the `override path`_ is specified, the `search paths`_ will be ignored.
+
+- For `csfle` `search paths`_ and the `override path`_, if the given path is a
+  relative path and the first path component is the literal string
+  "``$ORIGIN``", the "``$ORIGIN``" component will be replaced by the absolute
+  path to the directory containing the `libmongocrypt` library that is
+  performing the `csfle` search. This behavior mimics the ``$ORIGIN`` behavior
+  of the ``RUNPATH``/``RPATH`` properties of ELF executable files. This permits
+  bundling the `csfle` library along with `libmongocrypt` for creating portable
+  application distributions without relying on a externally/globally available
+  `csfle` library.
+
+  .. note:: No other ``RPATH``/``RUNPATH``-style substitutions are available.
+
+  .. _cse.csfle.dollar-system:
+
+- For `csfle` `search paths`_, if a search path string is "``$SYSTEM``", then
+  |---| instead of `libmongocrypt` searching for `csfle` in a directory named
+  "``SYSTEM``" |---| `libmongocrypt` will defer to the operating system's own
+  dynamic-library resolution mechanism when processing that search-path.
+
+- If a `search path`_ or `override path`_ is given as a relative path, that path
+  will be resolved relative to the working directory of the operating system
+  process.
+
+.. _search path:
+.. _search paths:
+.. _cse.csfle.search-paths:
+
+Adding Search Paths
+-------------------
+
+The driver MUST append csfle search paths to the `libmongocrypt_handle` in the
+following order:
+
+1. Append each search path in |opt-paths-prefer|.
+2. If `extraOptions.csfleDisableSystemLibrary <cse.extraoptions_>`__ *was not*
+   specified as ``true`` by the user, append the literal string "``$SYSTEM``" to
+   the path list. (`Refer <cse.csfle.dollar-system_>`_)
+3. Append each search path in |opt-paths-fallback|.
+
+.. note::
+
+   If |opt-path-override| is set on a `libmongocrypt_handle`, then these search
+   paths will have no effect.
+
+.. note::
+
+   If no |opt-paths-prefer| was specified or is an empty array, AND if no
+   |opt-paths-fallback| was specified or is an empty array, AND
+   `extraOptions.csfleDisableSystemLibrary <cse.extraOptions_>`__ was specified
+   as ``true``, the result should be that no csfle search paths will be added to
+   the `libmongocrypt_handle`, therefore libmongocrypt *will not* search for the
+   csfle library.
```

Comment:
This means that the default behavior, with no `extraOptions` passed, will be that the driver passes only `$SYSTEM` to libmongocrypt using `mongocrypt_setopt_append_csfle_search_path()`, libmongocrypt will attempt to load the shared library, and if it fails to do so, the driver will attempt to spawn mongocryptd.

Is that correct? It means that an application can load either the shared library or mongocryptd through the same application code, only depending on whether they are using a version of the driver that includes this feature or not.

Reply:
This is correct according to my intentions when writing, but raises a good qeustion: I presumed "load and use csfle if possible" was the default so as to hide the command-marking process, but that may be surprising to users who currently expect `mongocrypd` to appear, even though we try to keep `mongocryptd`/`csfle` "hidden" as best we can.